panic: runtime error: invalid memory address or nil pointer dereference

Hi,

Whenever I run the spreed backend container, it crashes a few seconds later. I am running it on a Raspberry Pi 4 (arm32v6).

Here are its logs:
```
2021-01-05 04:00:50.282136 I | main.go:130: Starting up version 97a604915479b16d6faa996e5218fb38341db751/go1.13.15 as pid 1
2021-01-05 04:00:50.283188 I | main.go:139: Using a maximum of 4 CPUs
2021-01-05 04:00:50.311139 I | natsclient.go:116: Connection established to nats://nextcloud-nats:4222 (NBJ625S6UI2Y6DMZUN6K33H2GGJ2EN2L2ELELLP4UHFLXHBQMS5LCZF2)
2021-01-05 04:00:50.311628 I | backend_configuration.go:77: Backend nextcloud added for https://MYDOMAIN
2021-01-05 04:00:50.312374 I | hub.go:177: Using a maximum of 8 concurrent backend connections per host
2021-01-05 04:00:50.312499 I | hub.go:184: Using a timeout of 10s for backend connections
2021-01-05 04:00:50.312667 I | hub.go:270: Not using GeoIP database
2021-01-05 04:00:50.325247 I | mcu_janus.go:289: Connected to Janus WebRTC Server 0.9.2 by Meetecho s.r.l.
2021-01-05 04:00:50.325315 I | mcu_janus.go:293: Found JANUS VideoRoom plugin 0.0.9 by Meetecho s.r.l.
2021-01-05 04:00:50.325350 I | mcu_janus.go:299: Data channels are supported
2021-01-05 04:00:50.325382 I | mcu_janus.go:305: Full-Trickle is enabled
2021-01-05 04:00:50.325416 I | mcu_janus.go:308: Maximum bandwidth 1048576 bits/sec per publishing stream
2021-01-05 04:00:50.325449 I | mcu_janus.go:309: Maximum bandwidth 2097152 bits/sec per screensharing stream
2021-01-05 04:00:50.333802 I | mcu_janus.go:315: Created Janus session 8792596263727263
2021-01-05 04:00:50.341194 I | mcu_janus.go:322: Created Janus handle 6470059305944818
2021-01-05 04:00:50.341272 I | main.go:226: Using janus MCU
2021-01-05 04:00:50.341334 I | hub.go:357: Using a timeout of 10s for MCU requests
2021-01-05 04:00:50.342092 I | backend_server.go:94: Using configured TURN API key
2021-01-05 04:00:50.342372 I | backend_server.go:95: Using configured shared TURN secret
2021-01-05 04:00:50.342428 I | backend_server.go:97: Adding "turn:nextcloud-coturn:9991?transport=udp" as TURN server
2021-01-05 04:00:50.342463 I | backend_server.go:97: Adding "turn:nextcloud-coturn:9991?transport=tcp" as TURN server
2021-01-05 04:00:50.342509 I | backend_server.go:109: Only allowing access to the stats endpoing from 127.0.0.1,nextcloud-spreedbackend
2021-01-05 04:00:50.345141 I | main.go:302: Listening on 0.0.0.0:8080
2021-01-05 04:00:59.529493 I | client.go:251: Client from MYIP has RTT of 62 ms (62.685244ms)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12670]

goroutine 43 [running]:
runtime/internal/atomic.goXadd64(0x1614084, 0x1, 0x0, 0x159bb5c, 0x1485660)
        /usr/local/go/src/runtime/internal/atomic/atomic_arm.go:103 +0x1c
signaling.(*Hub).processRegister(0x1614000, 0x15d4640, 0x14943c0, 0x15c2160, 0x1485620)
        /workdir/src/signaling/hub.go:677 +0xe8
signaling.(*Hub).processHelloClient(0x1614000, 0x15d4640, 0x14943c0)
        /workdir/src/signaling/hub.go:906 +0x234
signaling.(*Hub).processHello(0x1614000, 0x15d4640, 0x14943c0)
        /workdir/src/signaling/hub.go:873 +0xa90
signaling.(*Hub).processMessage(0x1614000, 0x15d4640, 0x168a000, 0x129, 0x600)
        /workdir/src/signaling/hub.go:780 +0x9b0
signaling.(*Client).ReadPump(0x15d4640)
        /workdir/src/signaling/client.go:297 +0x430
created by signaling.(*Hub).serveWs
        /workdir/src/signaling/hub.go:1899 +0x24c
```

Does anyone have any idea what's happening and how to fix it ?

Many thanks to everyone who tries to help 👍

Thanks for trying, but unfortunately, even after deleting my images , recloning your repo, and building the images, the server still crashes as soon as I click on a Nextcloud Talk conversation with the High-Performance Backend defined in settings. :/

The logs look similar to the ones I posted last time.
```
2021-01-06 14:13:05.124882 I | main.go:130: Starting up version 464a683e53faea01682301b9d1e22dfd957170fc/go1.13.15 as pid 1
2021-01-06 14:13:05.125539 I | main.go:139: Using a maximum of 4 CPUs
2021-01-06 14:13:05.132989 I | natsclient.go:116: Connection established to nats://nextcloud-nats:4222 (ND4TVHMXG23ENYAXEB2C3HKHKNSGH6QD6HHZ7Y3OLSDKM2VBLKSFIKHB)
2021-01-06 14:13:05.133285 I | backend_configuration.go:77: Backend nextcloud added for https://MY_DOMAIN
2021-01-06 14:13:05.133380 I | hub.go:177: Using a maximum of 8 concurrent backend connections per host
2021-01-06 14:13:05.133482 I | hub.go:184: Using a timeout of 10s for backend connections
2021-01-06 14:13:05.133600 I | hub.go:270: Not using GeoIP database
2021-01-06 14:13:05.148206 I | mcu_janus.go:289: Connected to Janus WebRTC Server 0.9.2 by Meetecho s.r.l.
2021-01-06 14:13:05.148311 I | mcu_janus.go:293: Found JANUS VideoRoom plugin 0.0.9 by Meetecho s.r.l.
2021-01-06 14:13:05.148370 I | mcu_janus.go:299: Data channels are supported
2021-01-06 14:13:05.148456 I | mcu_janus.go:305: Full-Trickle is enabled
2021-01-06 14:13:05.148524 I | mcu_janus.go:308: Maximum bandwidth 1048576 bits/sec per publishing stream
2021-01-06 14:13:05.148587 I | mcu_janus.go:309: Maximum bandwidth 2097152 bits/sec per screensharing stream
2021-01-06 14:13:05.150347 I | mcu_janus.go:315: Created Janus session 1647349674304042
2021-01-06 14:13:05.155010 I | mcu_janus.go:322: Created Janus handle 8223569116807938
2021-01-06 14:13:05.156024 I | main.go:226: Using janus MCU
2021-01-06 14:13:05.156154 I | hub.go:357: Using a timeout of 10s for MCU requests
2021-01-06 14:13:05.156336 I | backend_server.go:94: Using configured TURN API key
2021-01-06 14:13:05.156427 I | backend_server.go:95: Using configured shared TURN secret
2021-01-06 14:13:05.156512 I | backend_server.go:97: Adding "turn:nextcloud-coturn:3478?transport=udp" as TURN server
2021-01-06 14:13:05.156595 I | backend_server.go:97: Adding "turn:nextcloud-coturn:3478?transport=tcp" as TURN server
2021-01-06 14:13:05.156709 I | backend_server.go:109: Only allowing access to the stats endpoing from 127.0.0.1,nextcloud-spreedbackend
2021-01-06 14:13:05.157875 I | main.go:302: Listening on 0.0.0.0:8080
2021-01-06 14:20:02.209345 I | client.go:251: Client from MY_IP_ADDRESS has RTT of 69 ms (69.885763ms)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12670]

goroutine 33 [running]:
runtime/internal/atomic.goXadd64(0x164a084, 0x1, 0x0, 0x15d1b5c, 0x15498a0)
        /usr/local/go/src/runtime/internal/atomic/atomic_arm.go:103 +0x1c
signaling.(*Hub).processRegister(0x164a000, 0x14a6dc0, 0x14ed770, 0x15da180, 0x1549860)
        /workdir/src/signaling/hub.go:677 +0xe8
signaling.(*Hub).processHelloClient(0x164a000, 0x14a6dc0, 0x14ed770)
        /workdir/src/signaling/hub.go:906 +0x234
signaling.(*Hub).processHello(0x164a000, 0x14a6dc0, 0x14ed770)
        /workdir/src/signaling/hub.go:873 +0xa90
signaling.(*Hub).processMessage(0x164a000, 0x14a6dc0, 0x166d200, 0x129, 0x600)
        /workdir/src/signaling/hub.go:780 +0x9b0
signaling.(*Client).ReadPump(0x14a6dc0)
        /workdir/src/signaling/client.go:297 +0x430
created by signaling.(*Hub).serveWs
        /workdir/src/signaling/hub.go:1899 +0x24c
```

--- a/src/proxy/proxy_server.go
+++ b/src/proxy/proxy_server.go
@@ -81,6 +81,9 @@ var (
 )
 
 type ProxyServer struct {
+	// 64-bit members that are accessed atomically must be 64-bit aligned.
+	load int64
+
 	version string
 	country string
 
@@ -88,7 +91,6 @@ type ProxyServer struct {
 	nats    signaling.NatsClient
 	mcu     signaling.Mcu
 	stopped uint32
-	load    int64
 
 	shutdownChannel   chan bool
 	shutdownScheduled uint32

--- a/src/proxy/proxy_session.go
+++ b/src/proxy/proxy_session.go
@@ -37,10 +37,12 @@ const (
 )
 
 type ProxySession struct {
-	proxy    *ProxyServer
-	id       string
-	sid      uint64
+	// 64-bit members that are accessed atomically must be 64-bit aligned.
 	lastUsed int64
+
+	proxy *ProxyServer
+	id    string
+	sid   uint64
 
 	clientLock      sync.Mutex
 	client          *ProxyClient

--- a/src/signaling/hub.go
+++ b/src/signaling/hub.go
@@ -96,6 +96,9 @@ const (
 )
 
 type Hub struct {
+	// 64-bit members that are accessed atomically must be 64-bit aligned.
+	sid uint64
+
 	nats         NatsClient
 	upgrader     websocket.Upgrader
 	cookie       *securecookie.SecureCookie
@@ -113,7 +116,6 @@ type Hub struct {
 	mu sync.RWMutex
 	ru sync.RWMutex
 
-	sid      uint64
 	clients  map[uint64]*Client
 	sessions map[uint64]Session
 	rooms    map[string]*Room

--- a/src/signaling/mcu_janus.go
+++ b/src/signaling/mcu_janus.go
@@ -136,6 +136,9 @@ type clientInterface interface {
 }
 
 type mcuJanus struct {
+	// 64-bit members that are accessed atomically must be 64-bit aligned.
+	clientId uint64
+
 	url  string
 	mu   sync.Mutex
 	nats NatsClient
@@ -150,7 +153,6 @@ type mcuJanus struct {
 
 	closeChan chan bool
 
-	clientId  uint64
 	muClients sync.Mutex
 	clients   map[clientInterface]bool
 

--- a/src/signaling/mcu_proxy.go
+++ b/src/signaling/mcu_proxy.go
@@ -259,6 +259,11 @@ func (s *mcuProxySubscriber) ProcessEvent(msg *EventProxyServerMessage) {
 }
 
 type mcuProxyConnection struct {
+	// 64-bit members that are accessed atomically must be 64-bit aligned.
+	reconnectInterval int64
+	msgId             int64
+	load              int64
+
 	proxy  *mcuProxy
 	rawUrl string
 	url    *url.URL
@@ -270,15 +275,12 @@ type mcuProxyConnection struct {
 	conn       *websocket.Conn
 
 	connectedSince    time.Time
-	reconnectInterval int64
 	reconnectTimer    *time.Timer
 	shutdownScheduled uint32
 	closeScheduled    uint32
 
-	msgId      int64
 	helloMsgId string
 	sessionId  string
-	load       int64
 	country    atomic.Value
 
 	callbacks map[string]func(*ProxyServerMessage)
@@ -959,6 +961,10 @@ func (c *mcuProxyConnection) newSubscriber(ctx context.Context, listener McuList
 }
 
 type mcuProxy struct {
+	// 64-bit members that are accessed atomically must be 64-bit aligned.
+	connRequests int64
+	nextSort     int64
+
 	tokenId  string
 	tokenKey *rsa.PrivateKey
 
@@ -972,8 +978,6 @@ type mcuProxy struct {
 	connections    []*mcuProxyConnection
 	connectionsMap map[string]*mcuProxyConnection
 	connectionsMu  sync.RWMutex
-	connRequests   int64
-	nextSort       int64
 	proxyTimeout   time.Duration
 
 	mu         sync.RWMutex

